dataclasses.py

```diff
@@ -528,13 +529,36 @@ class C:
 
       c = C(1, 2)
       assert asdict(c) == {'x': 1, 'y': 2}
+
+    If given, 'dict_factory' will be used instead of built-in dict.
+    The function applies recursively to field values that are
+    dataclass instances. This will also look into built-in containers: tuples,
+    lists, and dicts.
     """
-    if isinstance(obj, type):
-        raise ValueError("asdict() should be called on dataclass instances, not classes")
-    return {name: getattr(obj, name) for name in fields(obj)}
+    if isinstance(obj, type) or not hasattr(obj, _MARKER):
+        raise TypeError("asdict() should be called on dataclass instances")
+    return _asdict_inner(obj, dict_factory)
+
+def _asdict_inner(obj, dict_factory):
+    if not hasattr(obj, _MARKER):
+        return obj
+    result = []
+    for name in fields(obj):
+        value = getattr(obj, name)
+        if hasattr(value, _MARKER) and not isinstance(value, type):
+            value = _asdict_inner(value, dict_factory)
+        elif isinstance(value, (list, tuple)):
+            value = type(value)(_asdict_inner(v, dict_factory) for v in value)
+        elif isinstance(value, dict):
+            value = type(value)((k, _asdict_inner(v, dict_factory))
```

Comment:
I guess it's okay that we don't call the helper on the key, here.
```
@dataclass(frozen=True)
class Key:
    i: int

@dataclass
class Value:
    i: int

@dataclass
class A:
    d: Dict[Key, Value]

a = A({Key(1): Value(2)})
print(asdict(a))
```
Produces: `{'d': {Key(i=1): {'i': 2}}}`. We can't recursively turn the key into a dict, since that's not hashable. But in the `astuple()` case we could produce `({(1,): (2,)},)` instead of `({Key(i=1): (2,)},)`. Should we keep the consistency with `asdict()`? Or maybe we do want to call the helper in the `asdict()` case, just to turn this into an error (and to get the deepcopy on the key, but it should be immutable, anyway).

Thoughts?

Other than that, this looks complete.

Reply:
For what it's worth, attrs gives a TypeError: unhashable in this case:
```
import attr

@attr.s(frozen=True)
class Key:
    key = attr.ib()

@attr.s
class Value:
    value = attr.ib()

@attr.s
class A:
    d = attr.ib()

a = A({Key(0): Value(1)})
print(a)
print(attr.asdict(a))
```
Produces:
```
A(d={Key(key=0): Value(value=1)})
Traceback (most recent call last):
  File "testattrs.py", line 17, in <module>
    print(attr.asdict(a))
  File "/cygdrive/c/home/eric/local/venvs/attrs/lib/python3.6/site-packages/attr/_funcs.py", line 62, in asdict
    for kk, vv in iteritems(v))
TypeError: unhashable type: 'dict'
```
Given that, I think we should call the helper function on the keys as well as the values.